### PR TITLE
Store environment when creating new catalog entries 

### LINKED ISSUE #2523: Registering subhosts when creating new catalog entries advances the pseudo random number generator
For creating a new catalog entry in MLContext, the existing pattern is to register subhosts i.e. create a new Host out of the environment. This advances the pseudo random number generator. 

https://github.com/dotnet/machinelearning/blob/5c442a96edc3ac7a03bc10ef668310647f16db9a/src/Microsoft.ML.Data/TrainCatalog.cs#L195-L200

This means that if we add a new catalog entry in MLContext, the baseline tests would start failing.  (In fact this is the exact behaviour observed in PR #2390 when we try to add a AnomalyDetectionCatalog entry to MLContext)

Instead of *forking* the environment we can just *store* the environment passed in.

@TomFinley

### PR DESCRIPTION
**Pull request by abgoswam**
Fixes #2523 

The following changes are being made:

1. Just _store_ the environment passed in , instead of creating subhosts
2. Updated baselines to make tests pass

### CODE REVIEW

File: test/BaselineOutput/SingleDebug/Rff/featurized.tsv [hunk 1]
```diff
@@ -6,7 +6,7 @@
 #@   col=RffVectorFloat:R4:9-14
 #@ }
 15	8:Label
-5	1	1	1	2	1	3	1	0	-0.157029659	-0.555585265	0.490177631	-0.305056125	0.35670203	-0.453979075
-5	4	4	5	7	10	3	2	0	-0.375955045	0.43816793	-0.5670244	0.108704455	-0.271485656	-0.5095379
-3	1	1	1	2	2	3	1	0	-0.08380841	-0.571235061	0.4856296	-0.312245429	0.389987826	-0.4257262
-6	8	8	1	3	4	3	7	0	-0.2813567	0.504154444	-0.266616732	-0.512102365	-0.5723418	-0.07588247
+5	1	1	1	2	1	3	1	0	0.3418174	-0.465289354	0.247575819	-0.5215741	0.576975167	-0.02080728
```
Comment (TomFinley) [inline test/BaselineOutput/SingleDebug/Rff/featurized.tsv]: I think you'll also need to change this in the release @abgoswam. You could also delete it, and move it to common.

File: src/Microsoft.ML.Data/TrainCatalog.cs [hunk 2]
```diff
@@ -20,7 +20,7 @@ namespace Microsoft.ML
     /// </summary>
     public abstract class TrainCatalogBase
     {
-        protected internal readonly IHost Host;
+        protected internal IHostEnvironment Host { get; set; }
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/TrainCatalog.cs]: >protected internal [](start = 8, length = 18)

`private protected` FYI

File: src/Microsoft.ML.Data/TrainCatalog.cs [hunk 3]
```diff
@@ -20,7 +20,7 @@ namespace Microsoft.ML
...
 
         [BestFriend]
         internal IHostEnvironment Environment => Host;
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/TrainCatalog.cs]: I don't see the point of having both. Can we just this, as an auto-implemented property?

File: src/Microsoft.ML.Data/TrainCatalog.cs (hunk 2 above)
Comment (TomFinley) [inline src/Microsoft.ML.Data/TrainCatalog.cs]: >set [](start = 56, length = 3)

Why do you think you need a set?

File: src/Microsoft.ML.Recommender/RecommenderCatalog.cs [hunk 4]
```diff
@@ -33,6 +33,7 @@ public sealed class RecommendationCatalog : TrainCatalogBase
         internal RecommendationCatalog(IHostEnvironment env)
             : base(env, nameof(RecommendationCatalog))
         {
+            Host = env;
```
Comment (TomFinley) [inline src/Microsoft.ML.Recommender/RecommenderCatalog.cs]: >Host [](start = 12, length = 4)

You're passing this in to the base constructor. Why are you setting it here?

Comment (abgoswam) [reply src/Microsoft.ML.Recommender/RecommenderCatalog.cs]: removed

---
In reply to: [256217269](https://github.com/dotnet/machinelearning/pull/2524#discussion_r256217269) [](ancestors = 256217269)

File: src/Microsoft.ML.Data/TrainCatalog.cs (hunk 3 above)
Comment (abgoswam) [reply src/Microsoft.ML.Data/TrainCatalog.cs]: did u mean  ".. just *use* this" ?  

Removed the `Host` property, and using `Environment` instead. 

---
In reply to: [256217044](https://github.com/dotnet/machinelearning/pull/2524#discussion_r256217044) [](ancestors = 256217044)

File: src/Microsoft.ML.Data/TrainCatalog.cs (hunk 2 above)
Comment (abgoswam) [reply src/Microsoft.ML.Data/TrainCatalog.cs]: removed

---
In reply to: [256217066](https://github.com/dotnet/machinelearning/pull/2524#discussion_r256217066) [](ancestors = 256217066)

Comment (abgoswam) [reply src/Microsoft.ML.Data/TrainCatalog.cs]: removed the Host property

---
In reply to: [256216915](https://github.com/dotnet/machinelearning/pull/2524#discussion_r256216915) [](ancestors = 256216915)

File: test/BaselineOutput/SingleDebug/Rff/featurized.tsv (hunk 1 above)
Comment (abgoswam) [reply test/BaselineOutput/SingleDebug/Rff/featurized.tsv]: Thanks. Done.

---
In reply to: [256215459](https://github.com/dotnet/machinelearning/pull/2524#discussion_r256215459) [](ancestors = 256215459)

**Comment by abgoswam** [comment]
Thanks folks for the reviews!

**Pull request merged**